### PR TITLE
make `ConstEvaluatable` more strict

### PR DESCRIPTION
**Pull request by lcnr**
relevant zulip discussion: https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/.60ConstEvaluatable.60.20generic.20functions/near/204125452

Let's see how much this impacts. Depending on how this goes this should probably be a future compat warning.

Short explanation: we currently forbid anonymous constants which depend on generic types, e.g. `[0; std::mem::size_of::<T>]` currently errors.

We previously checked this by evaluating the constant and returned an error if that failed. This however allows things like
```rust
const fn foo<T>() -> usize {
    if std::mem::size_of::<*mut T>() < 8 { // size of *mut T does not depend on T
        std::mem::size_of::<T>()
    } else {
        8
    }
}

fn test<T>() {
    let _ = [0; foo::<T>()];
}
```
which is a backwards compatibility hazard. This also has worrying interactions with mir optimizations (https://github.com/rust-lang/rust/pull/74491#issuecomment-661890421) and intrinsics (#74538).

r? @oli-obk @eddyb

### CODE REVIEW

File: src/librustc_middle/mir/mod.rs [hunk 1]
```diff
@@ -165,6 +165,19 @@ pub struct Body<'tcx> {
     /// FIXME(oli-obk): rewrite the promoted during promotion to eliminate the cell components.
     pub ignore_interior_mut_in_const_validation: bool,
 
+    /// Does this body use generic parameters. This is used for the `ConstEvaluatable` check.
+    ///
+    /// Note that this does not actually mean that this body is not computable right now,
+    /// the repeat count in the following example is polymorphic, but can still be evaluated
+    /// without knowing anything about the type parameter `T`.
```
Comment (lcnr) [inline src/librustc_middle/mir/mod.rs:172]:
```suggestion
    /// Note that this does not actually mean that this body is not computable right now.
    /// The repeat count in the following example is polymorphic, but can still be evaluated
    /// without knowing anything about the type parameter `T`.
```

**Comment by lcnr** [comment]
It looks like the crater queue is fairly full. @Mark-Simulacrum do we currently have the capacity to do another check run?

I also did some small changes here, so once again @bors try

**Comment by lcnr** [comment]
Well, will go ahead and add this PR into the crate queue. If we don't have the capacity for this we can just remove it again.

@craterbot check

**Comment by Mark-Simulacrum** [comment]
@craterbot retry seems to have been spurious

**Comment by lcnr** [comment]
@craterbot retry

**Comment by lcnr** [comment]
This behavior was introduced in #70452 which was merged on the 15.04.2020, meaning that this has been only stable for 2 stable versions. (since 1.43.0)

So hopefully there aren't many uses of this yet

**Comment by Mark-Simulacrum** [comment]
@craterbot retry

**Comment by lcnr** [comment]
It looks like there is one relevant regression here:

https://github.com/vlad9486/pq-newhope/blob/e9cdcb8828efeb38d6ed544870a582ca5e2fa543/src/poly/poly.rs#L314

**Comment by lcnr** [comment]
Converted the error to a future incompat lint. Let's see if we have to move `is_polymorphic` into a query.

@bors try @rust-timer queue

File: src/librustc_middle/mir/mod.rs [hunk 2]
```diff
@@ -165,6 +165,19 @@ pub struct Body<'tcx> {
...
+    /// Note that this does not actually mean that this body is not computable right now.
+    /// The repeat count in the following example is polymorphic, but can still be evaluated
+    /// without knowing anything about the type parameter `T`.
+    ///
+    /// ```rust
+    /// fn test<T>() {
+    ///     let _ = [0; std::mem::size_of::<*mut T>()];
+    /// }
+    /// ```
+    pub is_polymorphic: bool,
```
Comment (oli-obk) [inline src/librustc_middle/mir/mod.rs:179]: Maybe mention that optimizations should not touch this even if they make the body not polymorphic anymore.

File: src/librustc_trait_selection/traits/fulfill.rs [hunk 3]
```diff
@@ -458,16 +459,17 @@ impl<'a, 'b, 'tcx> ObligationProcessor for FulfillProcessor<'a, 'b, 'tcx> {
...
-                    ) {
-                        Ok(_) => ProcessResult::Changed(vec![]),
-                        Err(err) => ProcessResult::Error(CodeSelectionError(ConstEvalFailure(err))),
-                    }
+                        obligation.param_env,
+                        obligation.cause.span,
+                    )
+                    .map_or_else(
+                        |e| ProcessResult::Error(CodeSelectionError(ConstEvalFailure(e))),
+                        |()| ProcessResult::Changed(vec![]),
```
Comment (oli-obk) [inline src/librustc_trait_selection/traits/fulfill.rs:471]: idk, i like matches for these, but if you feel strongly about it, keep your change, if not, let's keep the status quo.

File: src/librustc_trait_selection/traits/select/mod.rs [hunk 4]
```diff
@@ -536,17 +537,18 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
...
-                ) {
-                    Ok(_) => Ok(EvaluatedToOk),
-                    Err(ErrorHandled::TooGeneric) => Ok(EvaluatedToAmbig),
-                    Err(_) => Ok(EvaluatedToErr),
-                }
+                    obligation.param_env,
+                    obligation.cause.span,
+                )
+                .map(|()| EvaluatedToOk)
+                .or_else(|e| match e {
```
Comment (oli-obk) [inline src/librustc_trait_selection/traits/select/mod.rs:548]: since you need to match anyway, you can also keep matching on the outer thing

**Comment by lcnr** [comment]
How is this a small perf win? :thinking:

**Comment by oli-obk** [comment]
Better codegen due to some code being behind a function that isn't getting inlined?

File: src/librustc_session/lint/builtin.rs [hunk 5]
```diff
@@ -549,6 +549,16 @@ declare_lint! {
     };
 }
 
+declare_lint! {
+    pub CONST_EVALUATABLE_UNCHECKED,
+    Warn,
+    "detects a generic constant is used in a type without a emitting a warning",
+    @future_incompatible = FutureIncompatibleInfo {
+        reference: "TODO",
```
Comment (oli-obk) [inline src/librustc_session/lint/builtin.rs:557]: needs a tracking issue

**Comment by oli-obk** [comment]
So... we're now in the bind where we either emit this future incompat lint, nudging ppl away from it (they have no way to get away from it other than completely rewriting their code), or we don't emit it until we have wf bounds (stable! :scream: ). I'm not happy with either solution, but it seems bad to emit future incompat lints that don't have a workaround right now but will have one in the future.

File: src/librustc_session/lint/builtin.rs (hunk 5 above)
Comment (lcnr) [reply src/librustc_session/lint/builtin.rs:557]: yeah, waiting on `t-lang` decision

**Comment by lcnr** [comment]
> they have no way to get away from it other than completely rewriting their code

Note that this does not change behavior (as the const expr has to not depend on it's generic parameters to work rn,
so it should always be possible to replace all generic params with dummies).

We also only allowed this in repeat expressions, so the following doesn't work:
```rust
fn foo<T>() -> [u8; std::mem::size_of::<*mut T>()] {
    todo!()
}
```

It might be interesting to look into associated constants for const wf, which would mean that we keep allowing
```rust
struct Foo<T>(T);

impl<T> Foo<T> {
    const ASSOC: usize = 17;
    
    fn test() -> [u8; Self::ASSOC] {
        todo!();
    }
}
```
which means that this PR will only forbid things like `[u8; std::mem::size_of::<*mut T>()]` which seems a lot less worrying to me.

**Comment by nikomatsakis** [comment]
Discussed in the rust-lang/lang meeting today:

* It seems like we should go forward, but we would like to have an Inside Rust blog post that explains what is going on and why.
* Also, is it possible to have a future-compatibility warning? If not, is it possible to have a special note in the error that directs people to an issue where they can learn more? (cc @lcnr and @oli-obk on that).

**Comment by lcnr** [comment]
Opened #76200 as a tracking issue and updated the PR description.

This is now a future compat lint for both associated constants and methods. @oli-obk afaict this is ready for merge now.

Will try to get a draft for the internals blog ready now.

File: src/test/ui/const_evaluatable/function-call.rs [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    if std::mem::size_of::<*mut T>() < 8 { // size of *mut T does not depend on T
+        std::mem::size_of::<T>()
+    } else {
+        8
+    }
+}
+
+fn test<T>() {
+    let _ = [0; foo::<T>()];
+    //~^ WARN cannot use constants which depend on generic parameters in types
```
Comment (oli-obk) [inline src/test/ui/const_evaluatable/function-call.rs:13]: you need to set the lint to warn again

**Comment by lcnr** [comment]
I wasn't able to reproduce this locally, but added a `#[warn(const_evaluatable_unchecked)]` anyways.

**Comment by oli-obk** [comment]
r=me with CI passing

**Comment by lcnr** [comment]
How can I reproduce this locally? I don't understand this error

**Comment by oli-obk** [comment]
The only thing I can come up with that could cause this difference is `--target=i586-unknown-linux-gnu`. You may be able to do a `--pass=check` test run of that specific test and check why it does this. I'm fine with just ignoring the test on that platform once we know why it's happening

**Comment by Mark-Simulacrum** [comment]
This was an [improvement](https://perf.rust-lang.org/compare.html?start=d92155bf6ae0b7d79fc83cbeeb0cc0c765353471&end=e2be5f568d1f60365b825530f5b5cb722460591b) of up to 2% on wf-projection-stress. Presumably not an expected result?

**Comment by lcnr** [comment]
We were afraid of causing a regression here and had similar results :thinking: Probably some inlining changes or something

but not really expected

**Pull request merged**